Tests/Common/Data/TradeBarConsolidatorTests.cs

```diff
@@ -130,11 +130,11 @@ public void OneMinuteAlwaysFiresEveryTimeOnMinueDataExceptFirstPoint()
 
             consolidator.Update(new TradeBar {Close = 2m, Time = reference.AddMinutes(1)});
             Assert.IsNotNull(consolidated);
-            Assert.AreEqual(2, consolidated.Close);
+            Assert.AreEqual(1, consolidated.Close);
 
             consolidator.Update(new TradeBar {Close = 3m, Time = reference.AddMinutes(2)});
             Assert.IsNotNull(consolidated);
-            Assert.AreEqual(3, consolidated.Close);
+            Assert.AreEqual(2, consolidated.Close);
```

Comment:
It would only be a time step behind when it is TimeSpan-only mode. In
TimeSpan-only mode, we wouldn't know when the time span is fulfilled until
it's past. This is because TradeBar only has a Time, not duration. For
example, if the TimeSpan is 3 minutes, and we received 3 trade bars of
10:00:00, 10:01:00, 10:02:00, we wouldn't know whether the next trade bar
is going to be over 3 minutes. The next bars could be 10:02:01, 10:02:15,
etc. for all we know, and they should all be included in the consolidation,
as they are still within 3 minutes. Until a bar of 10:03:00 or later comes
in, we cannot tell whether the time span of 3 minutes are fulfilled.

The special case is when TimeSpan is TimeSpan.Zero, which in reality
doesn't really exist! Even an instant is not zero. But let's say,
TimeSpan.Zero exists. That just means we want to capture every single bar,
which is equivalent to maxCount = 1. In which case, we don't need to delay
the consolidation.

I hope that makes sense.

On Tue, Mar 24, 2015 at 5:24 PM, Michael <notifications@github.com> wrote:

> In Tests/Common/Data/TradeBarConsolidatorTests.cs
> <https://github.com/QuantConnect/Lean/pull/59#discussion_r27005310>:
>
> >
> >              consolidator.Update(new TradeBar {Close = 3m, Time = reference.AddMinutes(2)});
> >              Assert.IsNotNull(consolidated);
> > -            Assert.AreEqual(3, consolidated.Close);
> > +            Assert.AreEqual(2, consolidated.Close);
>
> I'm not sure this is the behavior we want here. I agree that when it
> swallows the first point it isn't good, but it looks like now we're always
> one time unit in the past. Before this change I would be getting the newest
> data immediately, but now I'm a time step behind.
>
> I wonder if we can fix this.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/QuantConnect/Lean/pull/59/files#r27005310>.
>